### PR TITLE
Install and use conda on ortmodule CI pipelines

### PR DESCRIPTION
**Pull request by baijumeswani**
**Description**:

```ORTModule``` pipelines to use ```conda``` installation and run tests with ```conda``` ```python```. This helps with dynamically installing new python packages after docker build.

Two tasks that this will help with:
- Post build installation of python wheel (onnxruntime) to run tests (instead of using the build folder).
- Merging pipelines that require same build configuration but different testing packages (for example the ```orttraining-ortmodule``` and ```orttraining-linux-gpu-ci-pipeline```)

### CODE REVIEW

File: tools/ci_build/github/linux/docker/scripts/install_python_deps.sh [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        git clone https://github.com/ROCmSoftwarePlatform/DeepSpeed
+        cd DeepSpeed &&\
+          rm requirements/requirements-sparse_attn.txt &&\
+          ${PYTHON_EXE} setup.py bdist_wheel &&\
+          ${PYTHON_EXE} -m pip install dist/deepspeed*.whl &&\
+	      cd ..
+      fi
+    fi
+  fi
+fi
```
Comment (thiagocrepaldi) [inline tools/ci_build/github/linux/docker/scripts/install_python_deps.sh:76]: possibly add 

````bash
rm -rf /usr/include/google
rm -rf /usr/$LIBDIR/libproto*
```

File: tools/ci_build/github/linux/docker/scripts/install_os_deps.sh [hunk 2]
```diff
@@ -114,43 +89,16 @@ unzip gradle-6.3-bin.zip
 mv /tmp/src/gradle-6.3 /usr/local/gradle
 
 if ! [ -x "$(command -v protoc)" ]; then
-  source ${0/%install_deps\.sh/install_protobuf\.sh}
+  source ${0/%install_os_deps\.sh/install_protobuf\.sh}
 fi
 
 export ONNX_ML=1
 export CMAKE_ARGS="-DONNX_GEN_PB_TYPE_STUBS=OFF -DONNX_WERROR=OFF"
```
Comment (thiagocrepaldi) [inline tools/ci_build/github/linux/docker/scripts/install_os_deps.sh:96]: Possibly delete those guys in the final version

File: tools/ci_build/github/linux/docker/scripts/install_python_deps.sh [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+c) USE_CONDA=true;;
+esac
+done
+
+echo "Python version=$PYTHON_VER"
+
+DEVICE_TYPE=${DEVICE_TYPE:=Normal}
+
+if [[ $USE_CONDA = true ]]; then
+  # conda python version has already been installed.
```
Comment (thiagocrepaldi) [inline tools/ci_build/github/linux/docker/scripts/install_python_deps.sh:30]:
```suggestion
  # conda python version has already been installed by 
  #    tools/ci_build/github/linux/docker/Dockerfile.ubuntu_gpu_training.
```

File: tools/ci_build/github/linux/docker/Dockerfile.ubuntu_gpu_training [hunk 4]
```diff
@@ -26,3 +31,29 @@ ARG BUILD_UID=1000
...
+ARG MINICONDA_PREFIX=/home/$BUILD_USER/miniconda3
+RUN if [ "$USE_CONDA" = true ] ; \
+    then MINICONDA=miniconda.sh && \
+    wget --no-check-certificate --no-verbose https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O $MINICONDA && \
+    chmod a+x $MINICONDA && \
+    ./$MINICONDA -b -p $MINICONDA_PREFIX && \
+    rm ./$MINICONDA && \
+    $MINICONDA_PREFIX/bin/conda clean --yes --all && \
+    $MINICONDA_PREFIX/bin/conda install -y python=$PYTHON_VERSION && \
+    $MINICONDA_PREFIX/bin/conda install -c conda-forge -y protobuf ; \
```
Comment (thiagocrepaldi) [inline tools/ci_build/github/linux/docker/Dockerfile.ubuntu_gpu_training:44]: If pipeline fail, test without this protobuf install

Comment (snnn) [reply tools/ci_build/github/linux/docker/Dockerfile.ubuntu_gpu_training:44]: Could you please remove the "--no-check-certificate" ? It looks unsafe

Comment (snnn) [reply tools/ci_build/github/linux/docker/Dockerfile.ubuntu_gpu_training:44]: Is it possible to fix the version of Miniconda3?

File: tools/ci_build/github/linux/docker/scripts/install_python_deps.sh (hunk 1 above)
Comment (baijumeswani) [reply tools/ci_build/github/linux/docker/scripts/install_python_deps.sh:76]: I am not sure if I can delete those without root access. I will need to do that in the main dockerfile.

File: tools/ci_build/github/linux/docker/Dockerfile.ubuntu_gpu_training (hunk 4 above)
Comment (baijumeswani) [reply tools/ci_build/github/linux/docker/Dockerfile.ubuntu_gpu_training:44]: ok, I'll do that.

File: tools/ci_build/github/linux/docker/scripts/install_python_deps.sh (hunk 1 above)
Comment (snnn) [reply tools/ci_build/github/linux/docker/scripts/install_python_deps.sh:77]: When you build the docker image, you are running all the scripts with root.

However, I still don't understand why you say you can't install new python packages after docker build. It should not require root access. New packages can be installed to the user's folder.  Unless you need to touch some preinstalled packages.

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml [hunk 5]
```diff
@@ -38,6 +39,8 @@ jobs:
     condition: succeededOrFailed()
 
     # Entry point for all ORTModule tests
+    # Test run not from inside the /build/RelWithDebInfo/ to ensure the
```
Comment (thiagocrepaldi) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml:42]:
```suggestion
    # IMPORTANT: Test CANNOT run from /build/RelWithDebInfo/ to ensure the
```

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml [hunk 6]
```diff
@@ -38,6 +39,8 @@ jobs:
...
+    # tests pick up the installed wheel package of onnxruntime
```
Comment (thiagocrepaldi) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml:43]:
```suggestion
```

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml (hunk 5 above)
Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml]: Updated the comment to reflect the deleted ```onnxruntime``` folder in the build directory.

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml (hunk 6 above)
Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml]: Same as above. Updated the comment to indicate the deletion of the onnxruntime folder in the build dir.

File: tools/ci_build/github/linux/docker/scripts/requirements.txt [hunk 7]
```diff
@@ -4,7 +4,7 @@ mypy
 pytest
 setuptools>=41.4.0
 wheel
-git+http://github.com/onnx/onnx.git@971632833036c576cf95499291f689f8bb3519e1#egg=onnx
+onnx==1.9.0
```
Comment (snnn) [inline tools/ci_build/github/linux/docker/scripts/requirements.txt:7]: Is this change needed? If yes, could you please apply the same change to https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/linux/docker/scripts/manylinux/requirements.txt ? We should use the same ONNX version across all the pipelines. Also, please help us update C:\src\onnxruntime\cmake\external\onnx to the same commit id.

**Review by liqunfu — COMMENTED** [review]
lgtm, before merge this PR please test the following :
https://aiinfra.visualstudio.com/Lotus/_build?definitionId=872
https://dev.azure.com/onnxruntime/onnxruntime/_build?definitionId=166
https://dev.azure.com/onnxruntime/onnxruntime/_build?definitionId=163
The last one will fail due to pkg already published on pypi which I am working on a fix.

File: tools/ci_build/github/linux/docker/scripts/requirements.txt (hunk 7 above)
Comment (baijumeswani) [reply tools/ci_build/github/linux/docker/scripts/requirements.txt:7]: I'll try running the entire CI without this change in the hopes that the change is not required. If that fails, I will update all the other files you have mentioned as well.

Comment (snnn) [reply tools/ci_build/github/linux/docker/scripts/requirements.txt:7]: Thank you

File: tools/ci_build/github/linux/docker/scripts/install_python_deps.sh [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+DEVICE_TYPE=${DEVICE_TYPE:=Normal}
+
+if [[ $USE_CONDA = true ]]; then
+  # conda python version has already been installed by
+  # tools/ci_build/github/linux/docker/Dockerfile.ubuntu_gpu_training.
+  # so, /home/onnxruntimedev/miniconda3/bin/python should point
+  # to the correct version of the python version
+   PYTHON_EXE="/home/onnxruntimedev/miniconda3/bin/python"
+elif [[ "$PYTHON_VER" = "3.5" && -d "/opt/python/cp35-cp35m"  ]]; then
```
Comment (snnn) [inline tools/ci_build/github/linux/docker/scripts/install_python_deps.sh:35]: You can remove python 3.5. We don't support it anymore.

**Comment by SherlockNoMad** [comment]
Please add labels to future PRs.

File: tools/ci_build/github/linux/docker/scripts/requirements.txt (hunk 7 above)
Comment (baijumeswani) [reply tools/ci_build/github/linux/docker/scripts/requirements.txt]: Leaving it unchanged seems to have worked. Thanks.

**Review by snnn — APPROVED** [review]

**Comment by baijumeswani** [comment]
Thanks for the review.

**Comment by thiagocrepaldi** [comment]
nice!
